### PR TITLE
DLA-Future

### PR DESCRIPTION
**Pull request by RMeli**
DLA-Future integration in CP2K, for internal review.

# Regression Tests

DLA-Future: `941f216809c39a5f078a0ed26f55effa3072de4f` (branch `v0.2.1-instrumented`)

CP2K: `cdc57b27adaaccfba2e7162e31fe5c5f49339103` (branch `dlaf`)

CP2K has been patched as follows to run the regression tests:
```diff
diff --git a/src/input_cp2k_global.F b/src/input_cp2k_global.F
index 7e1c8fc17..4820dfbfc 100644
--- a/src/input_cp2k_global.F
+++ b/src/input_cp2k_global.F
@@ -214,7 +214,7 @@ CONTAINS
                           "the DLAF library. The eigensolver from the ScaLAPACK library is used as fallback "// &
                           "for all smaller cases", &
                           usage="DLAF_NEIGVEC_MIN 512", &
-                          default_i_val=1024)
+                          default_i_val=3)
       CALL section_add_keyword(section, keyword)
       CALL keyword_release(keyword)
```

## DLA-Future CPU

Spack environment (concretized the 26th of October 2023):
```yaml
spack:
  specs:
  - mpich device=ch3 netmod=tcp
  - intel-mkl threads=openmp
  - cosma+scalapack+shared
  - dbcsr@develop +mpi_f08
  - cp2k@master +libxc +libint +cosma +spglib +dlaf build_system=cmake 
  packages:
    all:
      variants:
      - +mpi
      - ~cuda
      - ~rocm
  repos:
  - ~/git/my-spack/repos/cp2k-dlaf/
  - ~/git/DLA-Future/spack/
  develop:
    dla-future:
      path: ~/git/DLA-Future
      spec: dla-future@master +hdf5 +scalapack ~cuda ~rocm build_type=Debug ^pika@main
  view: false
  concretizer:
    unify: true
``` 

Compilation command:
```bash
cmake -DCP2K_SCALAPACK_VENDOR=MKL -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCP2K_USE_DLAF=ON -DCP2K_ENABLE_REGTESTS=ON -DCP2K_USE_MPI_F08=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
```

Regression test command:
```bash
./tests/do_regtest.py local psmp --mpiranks 2 --ompthreads 6 --num_gpus 0 --mpiexec "mpiexec --bind-to core:6 pika-bind --"
```

Test results:
```
------------------------------- Summary --------------------------------
Number of FAILED  tests 2
Number of WRONG   tests 0
Number of CORRECT tests 2979
Total number of   tests 2981

Summary: correct: 2979 / 2981; failed: 2; 39min
Status: FAILED

*************************** Testing ended ******************************
```

Two test failed:
* `LIBTEST/dbm_blocks_06.inp`: errors related to Intel MKL
* `QS/regtest-pao-2/H2O_pao_equi_hybrid.inp`
  * Does not fail when running on a single rank

Cat of a test showing that pika and DLA-Future are correctly used:
```
*******************************************************************************
MPI rank: 0
   0: PU L#2(P#1), Core L#1(P#1), Socket L#0(P#0), on pool "default"
   1: PU L#4(P#2), Core L#2(P#2), Socket L#0(P#0), on pool "default"
   2: PU L#6(P#3), Core L#3(P#3), Socket L#0(P#0), on pool "default"
   3: PU L#8(P#4), Core L#4(P#4), Socket L#0(P#0), on pool "default"
   4: PU L#10(P#5), Core L#5(P#5), Socket L#0(P#0), on pool "default"
   5: PU L#0(P#0), Core L#0(P#0), Socket L#0(P#0), on pool "mpi"
*******************************************************************************
MPI rank: 1
   0: PU L#14(P#7), Core L#7(P#9), Socket L#0(P#0), on pool "default"
   1: PU L#16(P#8), Core L#8(P#10), Socket L#0(P#0), on pool "default"
   2: PU L#18(P#9), Core L#9(P#11), Socket L#0(P#0), on pool "default"
   3: PU L#20(P#10), Core L#10(P#12), Socket L#0(P#0), on pool "default"
   4: PU L#22(P#11), Core L#11(P#13), Socket L#0(P#0), on pool "default"
   5: PU L#12(P#6), Core L#6(P#8), Socket L#0(P#0), on pool "mpi"

[...]

  **** **** ******  **  PROGRAM STARTED AT               2023-10-27 09:34:15.820
 ***** ** ***  *** **   PROGRAM STARTED ON                               mercury
 **    ****   ******    PROGRAM STARTED BY                                 rmeli
 ***** **    ** ** **   PROGRAM PROCESS ID                                 87051
  **** **  *******  **  PROGRAM STARTED IN /home/rmeli/git/cp2k/regtesting/local
                                           /psmp/TEST-local-psmp-2023-10-27_09-1
                                           9-26/QS/regtest-ps-implicit-2-1

 CP2K| version string:                 CP2K version 2023.2 (Development Version)
 CP2K| source code revision number:                                    cdc57b27a
 CP2K| cp2kflags: ndebug omp parallel mpi_f08 scalapack mkl dlaf                
 CP2K| is freely available from                            https://www.cp2k.org/
 CP2K| Program compiled at                                   2023-10-27 08:49:39
 CP2K| Program compiled on                                               mercury
 CP2K| Program compiled for                                               x86_64
 CP2K| Data directory path                            /home/rmeli/git/cp2k/data/
 CP2K| Input file name                                     H2O_dbl_cstr_otcg.inp

 GLOBAL| Force Environment number                                              1
 GLOBAL| Basis set file name                                        BASIS_MOLOPT
 GLOBAL| Potential file name                                      GTH_POTENTIALS
 GLOBAL| MM Potential file name                                     MM_POTENTIAL
 GLOBAL| Coordinate file name                                      __STD_INPUT__
 GLOBAL| Method name                                                        CP2K
 GLOBAL| Project name                                          H2O_dbl_cstr_otcg
 GLOBAL| Run type                                                         ENERGY
 GLOBAL| FFT library                                                       FFTW3
 GLOBAL| Diagonalization library                                            DLAF
 GLOBAL| DGEMM library                                                      BLAS
 GLOBAL| Minimum number of eigenvectors for DLAF usage                         3
 GLOBAL| Orthonormality check for eigenvectors                          DISABLED
 GLOBAL| Matrix multiplication library                                 ScaLAPACK
 GLOBAL| All-to-all communication in single precision                          F
 GLOBAL| FFTs using library dependent lengths                                  F
 GLOBAL| Grid backend                                                       AUTO
 GLOBAL| Global print level                                               MEDIUM
 GLOBAL| MPI I/O enabled                                                       T
 GLOBAL| Total number of message passing processes                             2
 GLOBAL| Number of threads for this process                                    6
 GLOBAL| This output is from process                                           0
 GLOBAL| Stack size for threads created by OpenMP (OMP_STACKSIZE)        default
 GLOBAL| CPUID                                                              1003
```

## ScaLAPACK CPU

Spack environment: same as DLA-Future CPU.

Compilation command:
```bash
cmake -DCP2K_SCALAPACK_VENDOR=MKL -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCP2K_USE_DLAF=OFF -DCP2K_ENABLE_REGTESTS=ON -DCP2K_USE_MPI_F08=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
```

Regression test command:
```bash
./tests/do_regtest.py local psmp --mpiranks 2 --ompthreads 6 --num_gpus 0 --mpiexec "mpiexec --bind-to core:6 pika-bind --"
```

Test results:
```
------------------------------- Summary --------------------------------
Number of FAILED  tests 2
Number of WRONG   tests 0
Number of CORRECT tests 2978
Total number of   tests 2980

Summary: correct: 2978 / 2980; failed: 2; 16min
Status: FAILED

*************************** Testing ended ******************************
```

Two test failed:
* `LIBTEST/dbm_blocks_06.inp`: errors related to Intel MKL
* `QS/regtest-pao-2/H2O_pao_equi_hybrid.inp`
  * Does not fail when running on a single rank

### CODE REVIEW

**Comment by RMeli** [comment]
```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.12             Driver Version: 535.104.12   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4070        On  | 00000000:01:00.0  On |                  N/A |
|  0%   37C    P8               9W / 200W |    796MiB / 12282MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

```

## DLA-Future GPU

Spack environment:

```yaml
spack:
  specs:
  - mpich device=ch3 netmod=tcp
  - intel-mkl threads=openmp
  - cosma+scalapack+shared
  - dbcsr@develop +mpi_f08
  - cp2k@master +libxc +libint +cosma +spglib +dlaf build_system=cmake
  packages:
    all:
      variants:
      - +mpi
      - +cuda
      - cuda_arch=89
  repos:
  - ~/git/my-spack/repos/cp2k-dlaf/
  - ~/git/DLA-Future/spack/
  develop:
    dla-future:
      path: ~/git/DLA-Future
      spec: dla-future@master +hdf5 +scalapack +cuda cuda_arch=89 build_type=Debug ^pika@git.22a508dc5fa68813af53b8b2a53ca2e3138c3c96=main
  view: false
  concretizer:
    unify: true
```

Compilation command:
```bash
cmake -DCP2K_SCALAPACK_VENDOR=MKL -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCP2K_USE_DLAF=OFF -DCP2K_ENABLE_REGTESTS=ON -DCP2K_USE_MPI_F08=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCP2K_USE_ACCEL=CUDA -DCP2K_WITH_GPU=A100 ..
```

Regression test command:
```bash
./tests/do_regtest.py local_cuda psmp --mpiranks 2 --ompthreads 6 --num_gpus 1 --mpiexec "mpiexec --bind-to core:6 pika-bind --"
```

Test results:
```
------------------------------- Summary --------------------------------
Number of FAILED  tests 1
Number of WRONG   tests 0
Number of CORRECT tests 2980
Total number of   tests 2981

Summary: correct: 2980 / 2981; failed: 1; 70min
Status: FAILED

*************************** Testing ended ******************************
```

One test failed:
* `xTB/regtest-stda-force/h2o_f14.inp`
  * mismatch between analytical and numerical forces

File: .gitignore [hunk 1]
```diff
@@ -103,6 +103,8 @@ compile_commands.json
 CTestTestfile.cmake
 _deps
 
+build*/
+
```
Comment (msimberg) [inline .gitignore]: Does this belong here? Might be worth a small separate PR to CP2K?

File: CMakeLists.txt [hunk 2]
```diff
@@ -111,6 +111,8 @@ option(CP2K_USE_STATIC_BLAS "Link against static version of BLAS/LAPACK" OFF)
 option(CP2K_USE_SPLA
        "Use SPLA offloading gemm feature to the GPU if it is beneficial. " OFF)
 option(CP2K_USE_METIS "enable metis library support" OFF)
+option(CP2K_USE_DLAF "Use DLA-Future Cholesky decomposition and eigensolver"
```
Comment (msimberg) [inline CMakeLists.txt]:
```suggestion
option(CP2K_USE_DLAF "Use DLA-Future for Cholesky decomposition and eigensolver"
```
?

File: CMakeLists.txt [hunk 3]
```diff
@@ -578,6 +580,10 @@ if(CP2K_USE_VORI)
   find_package(LibVORI REQUIRED)
 endif()
 
+if(CP2K_USE_DLAF)
+  find_package(DLAF REQUIRED)
+endif()
```
Comment (msimberg) [inline CMakeLists.txt:591]: I think `cp2kConfig.cmake` will need a corresponding change.

File: INSTALL.md [hunk 4]
```diff
@@ -253,7 +253,16 @@ that provides tools for the solution of dense linear systems and eigenvalue prob
 - Add `-D__CUSOLVERMP` to `DFLAGS`
 - Add `-lcusolverMp -lcusolver -lcal -lnvidia-ml` to `LIBS`
 
-### 2m. PEXSI (optional, low scaling SCF method)
+### 2m. DLA-Future (optional, experimental, improved performance for diagonalization on Nvidia and AMD GPUs)
+
+[DLA-Future](https://github.com/eth-cscs/DLA-Future) is a high-performance, distributed-memory, GPU-accelerated library that provides tools for the solution of eigenvalue problems, based on the [pika](https://pikacpp.org/) runtime.
```
Comment (msimberg) [inline INSTALL.md]: Not sure if this is required, but line wrapping? Looks like other lines are wrapped to something like 80 characters.

File: cmake/modules/FindGenericBLAS.cmake [hunk 5]
```diff
@@ -43,7 +43,6 @@ endif()
 
 # add target to link against
 if(GenericBLAS_FOUND)
-
```
Comment (msimberg) [inline cmake/modules/FindGenericBLAS.cmake:46]: I don't mind this, and I don't know if the CP2K devs care, but it doesn't really belong here.

File: src/environment.F [hunk 6]
```diff
@@ -1062,6 +1072,7 @@ SUBROUTINE diag_setup_library(globenv)
                      elpa_qr=globenv%elpa_qr, &
                      elpa_print=globenv%elpa_print, &
                      elpa_qr_unsafe=globenv%elpa_qr_unsafe, &
+                     dlaf_neigvec_min_input=globenv%dlaf_neigvec_min, &
```
Comment (msimberg) [inline src/environment.F:1083]: Does the `_input` suffix belong here?

File: src/fm/cp_dlaf_utils.F [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+   PUBLIC :: cp_dlaf_initialize, cp_dlaf_finalize
+   PUBLIC :: cp_dlaf_create_grid, cp_dlaf_free_grid
+
+CONTAINS
+
+! **************************************************************************************************
+!> \brief Initialize DLA-Future and pika runtime
+!> \author Rocco Meli
+!> \author Mikeal Simberg
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.F]:
```suggestion
!> \author Mikael Simberg
```

File: src/fm/cp_dlaf_utils.F [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      CPABORT("CP2K compiled without the DLA-Future library.")
+#endif
+
+      CALL timestop(handle)
+   END SUBROUTINE cp_dlaf_initialize
+
+! **************************************************************************************************
+!> \brief Finalize DLA-Future and pika runtime
+!> \author Rocco Meli
+!> \author Mikeal Simberg
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.F]:
```suggestion
!> \author Mikael Simberg
```

File: src/fm/cp_dlaf_utils.F [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+#endif
+
+      CALL timestop(handle)
+   END SUBROUTINE cp_dlaf_create_grid
+
+! **************************************************************************************************
+!> \brief Free DLA-Future grid corresponding to BLACS context
+!> \param blacs_context ...
+!> \author Rocco Meli
+!> \author Mikeal Simberg
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.F]:
```suggestion
!> \author Mikael Simberg
```

File: src/fm/cp_dlaf_utils.F [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+#endif
+
+      CALL timestop(handle)
+   END SUBROUTINE cp_dlaf_finalize
+
+! **************************************************************************************************
+!> \brief Create DLA-Future grid from BLACS context
+!> \param blacs_context ...
+!> \author Rocco Meli
+!> \author Mikeal Simberg
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.F]:
```suggestion
!> \author Mikael Simberg
```

File: src/fm/cp_dlaf_utils.c [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+/*----------------------------------------------------------------------------*/
+
+#if defined(__DLAF)
+
+#include <dlaf_c/eigensolver/eigensolver.h>
+#include <dlaf_c/factorization/cholesky.h>
+#include <dlaf_c/init.h>
+
+void dlaf_init() {
+  const char *pika_argv[] = {"cp2k", "--pika:print-bind"};
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.c]: Just a reminder that we can remove the `--pika:print-bind` once we have the environment variable in 0.20.0.

File: src/fm/cp_dlaf_utils.c [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+void dlaf_pdpotrf_wrapper(int uplo, int n, double *a, int ia, int ja,
+                          int desca[9], int *info) {
+  char _uplo = ' ';
+  if (uplo == 0) {
+    _uplo = 'L';
+  } else if (uplo == 1) {
+    _uplo = 'U';
+  }
```
Comment (msimberg) [inline src/fm/cp_dlaf_utils.c]: This shows my ignorance with fortran, scalapack, etc. but what is this doing here? From my naive understanding this is always called with `'U'`, so in what situations would you expect to get `0` or `1`?

File: src/fm/cp_fm_dlaf_api.F [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+!> \brief Cholesky factorization using DLA-Future
+!> \param uplo ...
+!> \param n Matrix size
+!> \param a Local matrix
+!> \param ia Row index of first row (has to be 1)
+!> \param ja Col index of first column ()
+!> \param desca ScaLAPACK matrix descriptor
+!> \param info 0 if factorization completed normally
+!> \author Rocco Meli
+!> \author Mikeal Simberg
```
Comment (msimberg) [inline src/fm/cp_fm_dlaf_api.F]:
```suggestion
!> \author Mikael Simberg
```

File: src/input_cp2k_global.F [hunk 14]
```diff
@@ -206,6 +209,15 @@ SUBROUTINE create_global_section(section)
       CALL section_add_subsection(section, print_key)
       CALL section_release(print_key)
 
+      CALL keyword_create(keyword, __LOCATION__, name="DLAF_NEIGVEC_MIN", &
+                          description="Minimum number of eigenvectors for the use of the eigensolver from "// &
+                          "the DLAF library. The eigensolver from the ScaLAPACK library is used as fallback "// &
```
Comment (msimberg) [inline src/input_cp2k_global.F]:
```suggestion
                          "the DLA-Future library. The eigensolver from the ScaLAPACK library is used as fallback "// &
```
?

**Review by msimberg — CHANGES_REQUESTED** [review]
Very nice, I only have some minor superficial comments. Can't say much about the fortran side of things anyway...

File: CMakeLists.txt (hunk 2 above)
Comment (RMeli) [reply CMakeLists.txt]: Mmmh... I agree that how I wrote it might sound a bit off. But with "for eigensolver" doesn't sound right either. Maybe
```suggestion
option(CP2K_USE_DLAF "Use DLA-Future for Cholesky decompositions and the solution of eigenvalue problems"
```

File: CMakeLists.txt (hunk 3 above)
Comment (RMeli) [reply CMakeLists.txt:591]: Good catch, thanks!

File: src/fm/cp_dlaf_utils.c (hunk 12 above)
Comment (RMeli) [reply src/fm/cp_dlaf_utils.c]: This ugly workaround stems from my Fortran ignorance: it did not manage to pass a `CHARACTER` to C, so I had to resort to numbers.

It is indeed always called with `'U'` for now, but since for the Cholesky decomposition DLA-Future supports both values of UPLO, I wanted to support both.

But given the ugliness of the workaround, probably it is not worth it...

**Comment by RMeli** [comment]
The `QS/regtest-pao-2/H2O_pao_equi_hybrid` failure seems random, and is obtained also with ScaLAPACK. I also get them with the `master` branch.

@mtaillefumier is unable to reproduce them.

File: src/CMakeLists.txt [hunk 15]
```diff
@@ -962,9 +962,11 @@ list(
   fm/cp_cfm_diag.F
   fm/cp_cfm_basic_linalg.F
   fm/cp_cfm_types.F
+  fm/cp_dlaf_utils.F
```
Comment (mtaillefumier) [inline src/CMakeLists.txt]: exclude it from the build if dlaf is disabled.

File: src/f77_interface.F [hunk 16]
```diff
@@ -26,6 +26,8 @@ MODULE f77_interface
    USE cell_methods,                    ONLY: init_cell
    USE cell_types,                      ONLY: cell_type
    USE cp2k_info,                       ONLY: get_runtime_info
+   USE cp_dlaf_utils,                   ONLY: cp_dlaf_finalize,&
```
Comment (mtaillefumier) [inline src/f77_interface.F]: all of this should de decorated with #ifdef statements....

File: src/fm/cp_fm_diag.F [hunk 17]
```diff
@@ -210,6 +224,13 @@ SUBROUTINE choose_eigv_solver(matrix, eigenvectors, eigenvalues, info)
             CALL cp_fm_diag_cusolver(matrix, eigenvectors, eigenvalues)
          END IF
 
+      ELSE IF (diag_type == FM_DIAG_TYPE_DLAF) THEN
```
Comment (mtaillefumier) [inline src/fm/cp_fm_diag.F:238]: #ifdef __DLAF missing

File: src/fm/cp_fm_diag.F [hunk 18]
```diff
@@ -250,6 +271,8 @@ SUBROUTINE check_diag(matrix, eigenvectors, nvec)
          diag_type_name = "ELPA"
       ELSE IF (diag_type == FM_DIAG_TYPE_CUSOLVER) THEN
          diag_type_name = "CUSOLVER"
+      ELSE IF (diag_type == FM_DIAG_TYPE_DLAF) THEN
```
Comment (mtaillefumier) [inline src/fm/cp_fm_diag.F:287]: same here

File: src/environment.F [hunk 19]
```diff
@@ -600,6 +602,8 @@ SUBROUTINE read_global_section(root_section, para_env, globenv)
          CALL cite_reference(Marek2014)
       CASE (FM_DIAG_TYPE_CUSOLVER)
          globenv%diag_library = "cuSOLVER"
+      CASE (FM_DIAG_TYPE_DLAF)
```
Comment (mtaillefumier) [inline src/environment.F:607]: ifdef statement needed here. CP2K should exit right away if dlaf support is turned off. The message should indicate that DLAF support is OFF explicitly.

File: src/fm/cp_blacs_types.F [hunk 20]
```diff
@@ -17,6 +17,8 @@
 ! **************************************************************************************************
 MODULE cp_blacs_types
 
+   USE cp_dlaf_utils,                   ONLY: cp_dlaf_create_grid,&
```
Comment (mtaillefumier) [inline src/fm/cp_blacs_types.F]: #ifdef statement needed

File: src/fm/cp_fm_diag.F [hunk 21]
```diff
@@ -138,6 +149,8 @@ SUBROUTINE diag_init(diag_lib, fallback_applied, elpa_kernel, elpa_neigvec_min_i
 #endif
       ELSE IF (diag_lib == "cuSOLVER") THEN
          diag_type = FM_DIAG_TYPE_CUSOLVER
+      ELSE IF (diag_lib == "DLAF") THEN
```
Comment (mtaillefumier) [inline src/fm/cp_fm_diag.F:156]: I would put this instead
```
ELSE IF (diag_lib == "DLAF") THEN
#ifdef __DLAF
diag_type = FM_DIAG_TYPE_DLAF
#else
CPABORT("Error in diag_init: CP2K was compiled without DLAF support")
#endif

```

File: CMakeLists.txt (hunk 2 above)
Comment (msimberg) [reply CMakeLists.txt]: Yeah, good point! I like that formulation as well. I was considering commenting with
```suggestion
option(CP2K_USE_DLAF "Enable DLA-Future support"
```
since that's what most of the other options say, but it probably doesn't hurt to be a bit more explicit here about what dlaf is used for!

File: src/fm/cp_dlaf_utils.c (hunk 12 above)
Comment (msimberg) [reply src/fm/cp_dlaf_utils.c]: Ok, thanks for the explanation! I don't really understand the issue properly so I can't say if there's another way :shrug: So from my perspective this is fine, I was curious since it looked a bit odd. Does @mtaillefumier have some more experience with this? Is there another way of handling this?

Comment (RMeli) [reply src/fm/cp_dlaf_utils.c]: @msimberg I have a small reproducer for the issue: [fortran_to_c_char](https://github.com/RMeli/fortran-playground/tree/main/fortran_to_c_char) (`CHARACTER(len=*)` behaves as expected, `CHARACTER` not really). Of course I might be doing something wrong, I'm not a Fortran expert. It might be down to different sizes of the types? Unfortunately I couldn't find much information about this (all information I could find was focused on how to pass `CHARACTER(len=*)` to C, which works as expected).

IIRC, using integers instead of characters might have been @mtaillefumier's suggestion (or something coming up in a discussion).

Anyways, it's ugly. Since we only call with `"U"` anyways, I'll follow your suggestion.

File: src/environment.F (hunk 6 above)
Comment (RMeli) [reply src/environment.F:1083]: Yes, `dlaf_neigvec_min_input` is the name of the input variable, which sets the `dlaf_neigvec_min` variable in the `cp_fm_diag` module.

https://github.com/RMeli/cp2k/blob/a76d27d83272c686d563b8d44cfe03a86a81645a/src/fm/cp_fm_diag.F#L128-L129

File: src/CMakeLists.txt (hunk 15 above)
Comment (RMeli) [reply src/CMakeLists.txt]: I agree in principle, but this is not what it is done with the other packages. Do you prefer to go down a different path or be consistent with other libraries? I tried to do the latter.

File: src/f77_interface.F (hunk 16 above)
Comment (RMeli) [reply src/f77_interface.F]: Same question as above: I followed what it is done with ELPA, where the functions still exist and can be called, but then result in `CPABORT`. Do you want to go to a different route, where there is no DLA-Future related code in CP2K if `-DCP2K_USE_DLAF=off`?

File: src/fm/cp_dlaf_utils.c (hunk 11 above)
Comment (RMeli) [reply src/fm/cp_dlaf_utils.c]: `PIKA_PRINT_BIND` is available since https://github.com/pika-org/pika/pull/828.  https://github.com/pika-org/pika/pull/834 add an helper script (from @msimberg) with added support to set this variable.

Removing hard coded `--pika:print-bind` since it is no longer necessary. `pika@0.20.0` will be released soon.

File: CMakeLists.txt (hunk 2 above)
Comment (RMeli) [reply CMakeLists.txt]: I went for the first option. We can maybe switch to the second one if we will use DLA-Future GEMM too.

File: src/CMakeLists.txt (hunk 15 above)
Comment (RMeli) [reply src/CMakeLists.txt]: DLAF-related files added conditionally.

**Comment by RMeli** [comment]
Thanks @msimberg and @mtaillefumier! I think I addressed all your concerns.

File: src/f77_interface.F [hunk 22]
```diff
@@ -18,126 +18,130 @@
...
-                                              string_table_deallocate
-   USE timings,                         ONLY: add_timer_env,&
-                                              get_timer_env,&
-                                              rm_timer_env,&
-                                              timer_env_release,&
-                                              timer_env_retain,&
-                                              timings_register_hooks
-   USE timings_types,                   ONLY: timer_env_type
-   USE virial_types,                    ONLY: virial_type
+   USE base_hooks, ONLY: cp_abort_hook, &
```
Comment (msimberg) [inline src/f77_interface.F:21]: Is the indentation change here intentional?

File: src/start/cp2k_runs.F [hunk 23]
```diff
@@ -7,133 +7,137 @@
...
-                                              print_format_html
-   USE rt_propagation,                  ONLY: rt_prop_setup
-   USE sirius_interface,                ONLY: cp_sirius_finalize,&
-                                              cp_sirius_init
-   USE swarm,                           ONLY: run_swarm
-   USE tamc_run,                        ONLY: qs_tamc
-   USE tmc_setup,                       ONLY: do_analyze_files,&
-                                              do_tmc
-   USE vibrational_analysis,            ONLY: vb_anal
+   USE atom, ONLY: atom_code
```
Comment (msimberg) [inline src/start/cp2k_runs.F:10]: Same here, is the indentation change intentional?

File: src/fm/cp_fm_dlaf_api.F [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
```
Comment (msimberg) [inline src/fm/cp_fm_dlaf_api.F]:
```suggestion
!> \author Mikael Simberg
```
sorry... :upside_down_face:

File: src/f77_interface.F (hunk 22 above)
Comment (RMeli) [reply src/f77_interface.F:21]: Ooof... Looks like another issue with the `make pretty` command when using `#ifdef`/`#if defined`.

File: src/fm/cp_fm_dlaf_api.F (hunk 24 above)
Comment (RMeli) [reply src/fm/cp_fm_dlaf_api.F]: Sorry! I should have grepped the typo...

File: src/f77_interface.F (hunk 22 above)
Comment (RMeli) [reply src/f77_interface.F:21]: Related to https://github.com/cp2k/cp2k/issues/3071

Comment (RMeli) [reply src/f77_interface.F:21]: Won't fix. Let's deal with this on CP2K's side.

File: src/start/cp2k_runs.F (hunk 23 above)
Comment (RMeli) [reply src/start/cp2k_runs.F:10]: Won't fix. Let's deal with this on CP2K's side.

**Review by msimberg — APPROVED** [review]

**Comment by RMeli** [comment]
The failure with the GPU test `xTB/regtest-stda-force/h2o_f14.inp` is likely spurious too. Re-running it with the latest changes works: 

```bash
./tests/do_regtest.py local_cuda psmp --mpiranks 1 --ompthreads 6 --num_gpus 1 --mpiexec "mpiexec --bind-to core:6 pika-bind --print-bind --" --restrictdir xTB/regtest-stda-force
```

```
>>> /home/rmeli/git/cp2k/regtesting/local_cuda/psmp/TEST-local_cuda-psmp-2023-10-30_17-04-38/xTB/regtest-stda-force
    h2o_f12.inp                                                                           -5.850528424           OK (   1.80 sec)
    h2o_f13.inp                                                                           -5.768600886           OK (   1.88 sec)
    h2o_f14.inp                                                                           -5.768600886           OK (   2.03 sec)
    h2o_f18.inp                                                                           -5.769593107           OK (   2.04 sec)
<<< /home/rmeli/git/cp2k/regtesting/local_cuda/psmp/TEST-local_cuda-psmp-2023-10-30_17-04-38/xTB/regtest-stda-force (1 of 1) done in 7.76 sec

------------------------------- Summary --------------------------------
Number of FAILED  tests 0
Number of WRONG   tests 0
Number of CORRECT tests 4
Total number of   tests 4

Summary: correct: 4 / 4; 0min
Status: OK

*************************** Testing ended ******************************

```

**Review by mtaillefumier — APPROVED** [review]
